src/core/ext/filters/client_channel/resolver/polling_resolver.cc

```diff
@@ -99,30 +104,41 @@ void PollingResolver::ShutdownLocked() {
     gpr_log(GPR_INFO, "[polling resolver %p] shutting down", this);
   }
   shutdown_ = true;
-  if (have_next_resolution_timer_) {
-    grpc_timer_cancel(&next_resolution_timer_);
+  if (next_resolution_timer_handle_.has_value()) {
+    if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+      gpr_log(GPR_INFO, "[polling resolver %p] cancel re-resolution timer",
+              this);
+    }
+    event_engine_->Cancel(*next_resolution_timer_handle_);
   }
   request_.reset();
 }
 
-void PollingResolver::OnNextResolution(void* arg, grpc_error_handle error) {
-  auto* self = static_cast<PollingResolver*>(arg);
-  self->work_serializer_->Run(
-      [self, error]() { self->OnNextResolutionLocked(error); }, DEBUG_LOCATION);
+void PollingResolver::ScheduleNextResolutionTimer(const Duration& timeout) {
+  next_resolution_timer_handle_ =
+      event_engine_->RunAfter(timeout, [self = Ref()]() mutable {
+        ApplicationCallbackExecCtx callback_exec_ctx;
+        ExecCtx exec_ctx;
+        auto* self_ptr = static_cast<PollingResolver*>(self.get());
+        self_ptr->work_serializer_->Run(
+            [self = std::move(self)]() {
+              auto* self_ptr = static_cast<PollingResolver*>(self.get());
```

Comment:
This `self_ptr` is unnecessary here, this lambda can simply be:

```
[self = std::move(self)]() mutable { self->OnNextResolutionLocked(); }
```

Reply:
Disregard my suggestion, `self` here is a `Resolver*`, not a `PollingResolver*`.

Reply:
This problem comes up a lot.  The fact that `Ref()` on a subclass of `Resolver` returns a pointer to the base class type instead of a pointer to the subclass is a problem, but I haven't had time to figure out a good solution.

Fortunately, there's another bug in `RefCountedPtr<>` that offsets the above bug to some degree, which is that it allows down-casting without an explicit cast, which it really shouldn't.  But the fact that it does allow this means that you can do something like this:

```
RefCountedPtr<PollingResolver> self = Ref();
channel_args_.GetObject<EventEngine>()->RunAfter(
    timeout, [self = std::move(self)]() mutable { ... });
```

Now, inside the lambda, `self` is the subclass type, not the base class type.

It would be good to find a way to fix both of those bugs.  The second bug should be fairly easy to fix, although I don't want to fix it until we also fix the first one. :)